langchain4j-pinecone/src/test/java/dev/langchain4j/store/embedding/pinecone/PineconeEmbeddingStoreWithMetadataIT.java

```diff
@@ -0,0 +1,130 @@
+            .dimension(embeddingModel.dimension())
+            .metadataTypeMap(toMetadataTypeMap())
+            .createIndex(true)
+            .build();
+
+    @Test
+    void should_add_embedding_with_segment_with_metadata() {
+
+        Metadata metadata = createMetadata();
+
+        TextSegment segment = TextSegment.from("hello", metadata);
+        Embedding embedding = embeddingModel().embed(segment.text()).content();
+
+        String id = embeddingStore().add(embedding, segment);
+        assertThat(id).isNotBlank();
+
+        {
+            // Not returned.
+            TextSegment altSegment = TextSegment.from("hello?");
+            Embedding altEmbedding = embeddingModel().embed(altSegment.text()).content();
+            embeddingStore().add(altEmbedding, altSegment);
+        }
+
+        awaitUntilPersisted();
+
+        List<EmbeddingMatch<TextSegment>> relevant = embeddingStore().findRelevant(embedding, 1);
+        assertThat(relevant).hasSize(1);
+
+        EmbeddingMatch<TextSegment> match = relevant.get(0);
+        assertThat(match.score()).isCloseTo(1, withPercentage(1));
+        assertThat(match.embeddingId()).isEqualTo(id);
+        assertThat(match.embedding()).isEqualTo(embedding);
+
+        assertThat(match.embedded().text()).isEqualTo(segment.text());
+
+        assertThat(match.embedded().metadata().getString("string_empty")).isEqualTo("");
+        assertThat(match.embedded().metadata().getString("string_space")).isEqualTo(" ");
+        assertThat(match.embedded().metadata().getString("string_abc")).isEqualTo("abc");
+
+        assertThat(match.embedded().metadata().getDouble("double_minus_1")).isEqualTo(-1d);
+        assertThat(match.embedded().metadata().getDouble("double_0")).isEqualTo(Double.MIN_VALUE);
+        assertThat(match.embedded().metadata().getDouble("double_1")).isEqualTo(1d);
+        assertThat(match.embedded().metadata().getDouble("double_123")).isEqualTo(1.23456789d);
+
+        // new API
+        assertThat(embeddingStore().search(EmbeddingSearchRequest.builder()
+                .queryEmbedding(embedding)
+                .maxResults(1)
+                .build()).matches()).isEqualTo(relevant);
+    }
+
+    protected EmbeddingStore<TextSegment> embeddingStore() {
+        return embeddingStore;
+    }
+
+    protected EmbeddingModel embeddingModel() {
+        return embeddingModel;
+    }
+
+    @SneakyThrows
+    protected void awaitUntilPersisted() {
+        Thread.sleep(5000);
+    }
+
+    Metadata createMetadata() {
+
+        Metadata metadata = new Metadata();
+
+        metadata.put("string_empty", "");
+        metadata.put("string_space", " ");
+        metadata.put("string_abc", "abc");
+
+        metadata.put("double_minus_1", -1d);
+        metadata.put("double_0", Double.MIN_VALUE);
+        metadata.put("double_1", 1d);
+        metadata.put("double_123", 1.23456789d);
+
+        return metadata;
+    }
+
+    Map<String, Value.KindCase> toMetadataTypeMap() {
```

Comment:
Do we have to force user to specify this manually? Can't we figure it out in runtime when adding embeddings to the store? We could check the actual java types of metadata entries

Reply:
It's a good idea. But how to deserialize it when search? We can't get its Java type from the result of search

Reply:
@Martin7-1 what type
`getInteger()`, `getDouble()` and other `get*` methods in `Metadata` can convert from `String` or `Number`